### PR TITLE
Extend AasxFileRepo and relationship approach

### PR DESCRIPTION
**Pull request by MichaelHoffmeisterFesto**
* navigation between AASX
* jump back to specific plugin
* add some helpers in AasxCsharpLib
* null-fix in AasxPluginDocumentShelf
* experimental interoperable relationships
   (right now, only used by external plugin)
* preliminary definitons for MTP
* finding of MtpFile in Mtp plugin
* dragging of File contents works again

### CODE REVIEW

File: src/AasxCsharpLibrary/AdminShellUtil.cs [hunk 1]
```diff
@@ -92,6 +92,78 @@ public static string ExtractPascalCasingLetters(string src)
...
+            for (j = 0; j < line.Length; j++)
+                if (!Char.IsWhiteSpace(line[j]))
+                    break;
+            return j;
+        }
+
+        /// <summary>
+        /// Used to re-reformat a C# here string, which is multiline string introduced by @" ... ";
+        /// </summary>
+        public static string[] CleanHereStringToArray(string here)
```
Comment (mristin) [inline src/AasxCsharpLibrary/AdminShellUtil.cs:109]: Doctest?

File: src/AasxCsharpLibrary/AdminShellUtil.cs [hunk 2]
```diff
@@ -92,6 +92,78 @@ public static string ExtractPascalCasingLetters(string src)
...
+                lines.Insert(0, firstLine);
+
+            // return
+            return lines.ToArray();
+        }
+
+        /// <summary>
+        /// Used to re-reformat a C# here string, which is multiline string introduced by @" ... ";
+        /// </summary>
+        public static string CleanHereStringWithNewlines(string here, string nl = null)
```
Comment (mristin) [inline src/AasxCsharpLibrary/AdminShellUtil.cs:157]: Doctest?

File: src/AasxPackageExplorer/MainWindow.xaml.cs [hunk 3]
```diff
@@ -1022,6 +1045,17 @@ private void ButtonHistory_ObjectRequested(object sender, VisualElementHistoryIt
                         }
                     }
 
+                    // special case - 2nd half
+                    if (searchPluginTag != null && veFocus is VisualElementSubmodelRef veSm
+                        && veSm.Members != null)
```
Comment (mristin) [inline src/AasxPackageExplorer/MainWindow.xaml.cs:1050]: `&&` on the previous line for consistency with the rest of the codebase?

File: src/AasxPackageExplorer/MainWindow.xaml.cs [hunk 4]
```diff
@@ -1531,8 +1565,9 @@ private void Window_Drop(object sender, DragEventArgs e)
 
         private void DragSource_PreviewMouseMove(object sender, MouseEventArgs e)
         {
-            if (e.LeftButton == MouseButtonState.Pressed && !isDragging && (Math.Abs(dragStartPoint.X) < 0.001 &&
-                Math.Abs(dragStartPoint.Y) < 0.001) && this.showContentPackageUri != null &&
+            // MIHO 2020-09-14: removed this from the check below
```
Comment (mristin) [inline src/AasxPackageExplorer/MainWindow.xaml.cs:1568]: Maybe explain why?

**Review by mristin — APPROVED** [review]
LGTM, please see the minor remarks!

File: src/AasxCsharpLibrary/AdminShellPackageEnv.cs [hunk 5]
```diff
@@ -1059,13 +1059,30 @@ public void Dispose()
...
             string temppath = System.IO.Path.GetTempFileName().Replace(".tmp", tempext);
+
+            // maybe modify tempfile name?
+            if (keepFilename)
+            {
+                var masterFn = System.IO.Path.GetFileNameWithoutExtension(packageUri);
+                var tmpDir = System.IO.Path.GetDirectoryName(temppath);
+                var tmpFnExt = System.IO.Path.GetFileName(temppath);
+
+                temppath = System.IO.Path.Combine(tmpDir, "" + masterFn + "_" + tmpFnExt);
```
[the remark this answers is not in the document]
Comment (MichaelHoffmeisterFesto) [reply src/AasxCsharpLibrary/AdminShellPackageEnv.cs:1082]: For me, string interpolation would add complexity in this particular case.
In any case, I want to use Path.Combine, as it is OS independent.

File: src/AasxCsharpLibrary/AdminShell.cs [hunk 6]
```diff
@@ -5032,30 +5102,23 @@ public BaseSubmodelElementWrapperCollection(SubmodelElementWrapperCollection oth
 
                     // dive into?
                     // TODO (MIHO, 2020-07-31): would be nice to use IEnumerateChildren for this ..
```
[the remark this answers is not in the document]
Comment (MichaelHoffmeisterFesto) [reply src/AasxCsharpLibrary/AdminShell.cs:5104]: Yes, still valid

File: src/AasxCsharpLibrary/AdminShellUtil.cs [hunk 7]
```diff
@@ -92,6 +92,78 @@ public static string ExtractPascalCasingLetters(string src)
             return res;
         }
 
+        public static int CountHeadingSpaces(string line)
```
[the remark this answers is not in the document]
Comment (MichaelHoffmeisterFesto) [reply src/AasxCsharpLibrary/AdminShellUtil.cs:95]: For me Prio 2, as this fucntion is only enganged by string constants, which are never changed.

File: src/AasxCsharpLibrary/AdminShellUtil.cs (hunk 1 above)
Comment (MichaelHoffmeisterFesto) [reply src/AasxCsharpLibrary/AdminShellUtil.cs:109]: For me Prio 2, as this fucntion is only enganged by string constants, which are never changed.

File: src/AasxCsharpLibrary/AdminShellUtil.cs (hunk 2 above)
Comment (MichaelHoffmeisterFesto) [reply src/AasxCsharpLibrary/AdminShellUtil.cs:157]: For me Prio 2, as this function is only enganged by string constants, which are never changed.

File: src/AasxPackageExplorer/MainWindow.xaml.cs (hunk 3 above)
Comment (MichaelHoffmeisterFesto) [reply src/AasxPackageExplorer/MainWindow.xaml.cs:1050]: Please explain; I use this formatting quite often.

File: src/AasxPackageExplorer/MainWindow.xaml.cs (hunk 4 above)
Comment (MichaelHoffmeisterFesto) [reply src/AasxPackageExplorer/MainWindow.xaml.cs:1568]: I am not aware of the reason, why this was introduced.

File: src/AasxCsharpLibrary/AdminShellPackageEnv.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mristin) [reply src/AasxCsharpLibrary/AdminShellPackageEnv.cs:1082]: Sorry, I wasn't clear. This is what I had in mind:

```
temppath = System.IO.Path.Combine(tmpDir, $"{masterFn}_{tmpFnExt}");
```

File: src/AasxPackageExplorer/MainWindow.xaml.cs (hunk 3 above)
Comment (mristin) [reply src/AasxPackageExplorer/MainWindow.xaml.cs:1050]: When I reformatted the code to 120 char line length, most of the code had logical operators at the end of the line.

**Pull request merged**